crypto/include/internal/sparse_array.h

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef HEADER_SPARSE_ARRAY_H
+# define HEADER_SPARSE_ARRAY_H
+
+# ifdef __cplusplus
+extern "C" {
+# endif
+
+# define SPARSE_ARRAY_OF(type) struct sparse_array_st_ ## type
+
+# define DEFINE_SPARSE_ARRAY_OF(type) \
+    SPARSE_ARRAY_OF(type); \
+    static ossl_inline SPARSE_ARRAY_OF(type) * \
+        sa_##type##_new(void) \
+    { \
+        return (SPARSE_ARRAY_OF(type) *)OPENSSL_SA_new(); \
+    } \
+    static ossl_inline void sa_##type##_free(SPARSE_ARRAY_OF(type) *sa) \
+    { \
+        OPENSSL_SA_free((SA *)sa); \
+    } \
+    static ossl_inline void sa_##type##_free_leaves(SPARSE_ARRAY_OF(type) *sa) \
+    { \
+        OPENSSL_SA_free_leaves((SA *)sa); \
+    } \
+    static ossl_inline size_t sa_##type##_num(const SPARSE_ARRAY_OF(type) *sa) \
+    { \
+        return OPENSSL_SA_num((SA *)sa); \
+    } \
+    static ossl_inline void sa_##type##_doall(SPARSE_ARRAY_OF(type) *sa, \
+                                              void (*leaf)(type *)) \
+    { \
+        OPENSSL_SA_doall((SA *)sa, (void (*)(void *))leaf); \
+    } \
+    static ossl_inline void sa_##type##_doall_arg(SPARSE_ARRAY_OF(type) *sa, \
+                                                  void (*leaf)(type *, void *),\
+                                                  void *arg) \
+    { \
+        OPENSSL_SA_doall_arg((SA *)sa, (void (*)(void *, void *))leaf, arg); \
+    } \
+    static ossl_inline type *sa_##type##_get(SPARSE_ARRAY_OF(type) *sa, \
+                                            size_t n) \
+    { \
+        return (type *)OPENSSL_SA_get((SA *)sa, n); \
+    } \
+    static ossl_inline int sa_##type##_set(SPARSE_ARRAY_OF(type) *sa, \
+                                            size_t n, type *val) \
+    { \
+        return OPENSSL_SA_set((SA *)sa, n, (void *)val); \
+    } \
+    SPARSE_ARRAY_OF(type)
+
+typedef struct sparse_array_st SA;
```

Comment:
This should be OPENSSL_SA not SA ... we are trying to keep our namespace clean and with a common prefix.
Don't dig the hole deeper :-)

Reply:
For an _internal_ structure whose name isn't intended to ever be exported??
I'll change.

Reply:
No "SA" is a structure that you use in all the OPENSSL_SA APIs - the "SA" name is public and it should be OPENSSL_SA to be consistent. It is visible in a *public* API.

Reply:
???? the structure isn't public and is never intended to be.  We should never have public calls that accept a pointer to SA.  The sparse array APIs are deliberately internal only.

Reply:
Changed now.

Reply:
You are adding APIs called OPENSSL_SA - the type they take should be an OPENSSL_SA.
They will go in a header file. They will be used by other parts of the library. I expect them to become public APIs - and they should be named correctly from the beginning. The datatype should match the prefix in the function names.